pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -90,117 +102,310 @@ func (c *ClusterMemberController) reconcileMembers(ctx context.Context, recorder
+func (c *ClusterMemberController) getEtcdPeerURLToAdd(ctx context.Context) (string, error) {
+	nodes, err := c.masterNodeLister.List(c.masterNodeSelector)
+	if err != nil {
+		return "", err
+	}
+
+	nodesMapToVotingMembers, err := c.allNodesMapToVotingMembers(nodes)
+	if err != nil {
+		return "", fmt.Errorf("failed to map nodes to voting members: %v", err)
+	}
+	// If all nodes already map to voting members then there are no nodes
+	// that need to be scaled-up or added to the membership
+	if nodesMapToVotingMembers {
+		return "", nil
 	}
 
-	etcdHost, err := c.getEtcdPeerHostToScale(podToAdd)
+	members, err := c.etcdClient.MemberList(ctx)
 	if err != nil {
-		return fmt.Errorf("could not get etcd peer host :%w", err)
+		return "", err
 	}
 
-	if machineForHostPendingDeletion, err := c.isMachinePendingDeletionFor(etcdHost); err != nil {
+	// aggregate errors so we don't block on the failure to add one node
+	var errs []error
+	for _, node := range nodes {
+		if node.DeletionTimestamp != nil {
+			klog.V(2).Infof("Ignoring node (%s) for new member addition as it's pending deletion", node.Name)
+			continue
+		}
+
+		peerURL, err := c.getPeerURLForNode(node)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to get peerURL for node: %v", err))
+			continue
+		}
+
+		// Ignore if the node's PeerURL is mapped to a member and already part of the quorum.
+		if isURLMappedToMember(peerURL, members) {
+			continue
+		}
+
+		// If the Machine API is unavailable then we can't evaluate the machine's
+		// deletion timestamp or deletion hooks so we scale it up without any considerations
+		// for vertical scaling
+		// This allows UPI clusters without a machine API to scale up
+		// Check machine API availability
+		isMachineAPIFunctional, err := c.machineAPIChecker.IsFunctional()
+		if err != nil {
+			return "", fmt.Errorf("failed to determine Machine API availability: %v", err)
+		}
+		if !isMachineAPIFunctional {
+			return peerURL, nil
+		}
+
+		// If the Machine API is functional we should ignore the machine
+		// if it is pending deletion to avoid re-adding previously removed members
+		internalIP, err := c.getInternalIPForNode(node)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to get internal IP for node: %v", err))
+			continue
+		}
+		machine, err := ceohelpers.FindMachineByNodeInternalIP(internalIP, c.masterMachineSelector, c.masterMachineLister)
+		if err != nil {
+			return "", fmt.Errorf("failed to get machine for node (%s): %v", node.Name, err)
+		}
+		if machine == nil {
+			klog.V(2).Infof("Ignoring node (%s) for scale-up: no Machine found referencing this node's internal IP (%v)", node.Name, internalIP)
+			continue
+		}
+		if machine.DeletionTimestamp != nil {
+			// Ignore this node since its member previously was, or is in the process of being removed
+			klog.V(4).Infof("Ignoring node (%s) for scale-up since its machine (%s) is pending deletion", node.Name, machine.Name)
```

Comment:
we should be consistent, in other places within this method we do use `V(2)`. I think it is ok, this is important stuff, not an everyday operation. We should have full visibility into why a member has been ignored.